modules/jsonrpc/jsonrpc.cpp

```diff
@@ -126,23 +126,33 @@ Variant JSONRPC::process_action(const Variant &p_action, bool p_process_arr_elem
 
 		if (methods.has(method)) {
 			Variant call_ret = methods[method].callv(args);
-			if (id.get_type() != Variant::NIL) {
-				ret = make_response(call_ret, id);
-			}
+			ret = make_response(call_ret, id);
 		} else {
 			ret = make_response_error(JSONRPC::METHOD_NOT_FOUND, "Method not found: " + method, id);
 		}
+
+		/// The Server MUST NOT reply to a Notification
+		if (is_notification) {
+			ret = Variant();
+		}
 	} else if (p_action.get_type() == Variant::ARRAY && p_process_arr_elements) {
 		Array arr = p_action;
-		int size = arr.size();
-		if (size) {
+		if (!arr.is_empty()) {
 			Array arr_ret;
-			for (int i = 0; i < size; i++) {
-				const Variant &var = arr.get(i);
-				arr_ret.push_back(process_action(var));
+			for (const Variant &var : arr) {
+				Variant res = process_action(var);
+
+				if (res.get_type() != Variant::NIL) {
+					arr_ret.push_back(res);
+				}
+			}
+
+			/// If there are no Response objects contained within the Response array as it is to be sent to the client, the server MUST NOT return an empty Array and should return nothing at all.
```

Comment:
```suggestion
			// If there are no response objects contained within the response array as it is to be sent to the client, the server MUST NOT return an empty array and should return nothing at all.
```
Similar changes suggested for other comments, our comment style is:
```
// This is a comment ending with a period.
```

If "Response" was a specific term that's usually capitalized in JSONRPC context, then it's fine to keep capitalized, but I see some discrepancy between the capitalization of "Array" for example, so I suspect it's more of a germanism :P

Reply:
All comments starting with `///` are directly copied from the JSON RPC specification (except for some omissions which are marked using `...` and IMO should stay as is. If there is another standard for such comments I'll apply it, but it needs to be possible to distinguish implementation notes `//` and the specification `///`.